### PR TITLE
 test: added redux, api and selector test cases for in context topics

### PR DESCRIPTION
**Pull request by sundasnoreen12**
### Description
Added test cases for redux store manipulation, API responses, and selectors
https://2u-internal.atlassian.net/browse/INF-725

#### How Has This Been Tested?
I tested my changes by running the command "npm run test [relative path of the file]

#### Post-merge Checklist

* [ ] Deploy the changes to prod after verifying on stage or ask **@openedx/edx-infinity** to do it. 
* [ ] 🎉 🙌 Celebrate! Thanks for your contribution.

### CODE REVIEW

**Review by mehaknasir — CHANGES_REQUESTED** [review]
Good work with the code structuring and details, can you please look into thee points I have added.

**Review by awais-ansari — CHANGES_REQUESTED** [review]
I'm searching for the best practice for test cases and I find out no one is using selectors directly in test cases. It is happing because selectors would be tested on component integration testing. 
https://www.xtivia.com/blog/best-practices-for-testing-a-react-redux-toolkit-app/
https://www.willowtreeapps.com/craft/best-practices-for-unit-testing-with-a-react-redux-approach

File: src/discussions/in-context-topics/data/__factories__/inContextTopics.factory.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  .sequence('question', ['questionCount'], (questionCount) => questionCount);
+
+Factory.define('archived-topics')
+  .attr('id', null, 'archived')
+  .option('courseId', null, 'course-v1:edX+DemoX+Demo_Course')
+  .attr('children', ['id', 'courseId'], (id, courseId) => {
+    Factory.reset('topic');
+    return Factory.buildList('topic', 2, null, {
+      topicPrefix: `${id}`, enabledInContext: false, topicNamePrefix: `${id}`, usageKey: `${courseId.replace('course-v1:', 'block-v1:')} +type@vertical+block@`,
+    });
```
Comment (awais-ansari) [inline src/discussions/in-context-topics/data/__factories__/inContextTopics.factory.js:60]: line width is greater than 120. Please split this line into multiple lines.

File: src/discussions/in-context-topics/data/__factories__/inContextTopics.factory.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  });
+
+Factory.define('sub-section')
+  .sequence('block_id', (idx) => `${idx}`)
+  .option('topicPrefix', null, '')
+  .sequence('id', ['topicPrefix'], (idx, topicPrefix) => `${topicPrefix}-topic-${idx}`)
+  .sequence('display-name', ['sectionPrefix'], (idx, sectionPrefix) => `Introduction ${sectionPrefix + idx}`)
+  .option('courseId', null, 'course-v1:edX+DemoX+Demo_Course')
+  .sequence('legacy_web_url', ['id', 'courseId'], (idx, id, courseId) => `${getApiBaseUrl}/courses/${courseId}/jump_to/block-v1:${id}?experience=legacy`)
+  .sequence('lms_web_url', ['id', 'courseId'], (idx, id, courseId) => `${getApiBaseUrl}/courses/${courseId}/jump_to/block-v1:${id}`)
```
Comment (awais-ansari) [inline src/discussions/in-context-topics/data/__factories__/inContextTopics.factory.js:23]: These lines have width greater than 120. split these lines into multiple lines.

File: src/discussions/in-context-topics/data/__factories__/inContextTopics.factory.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  .sequence('student_view_url', ['id', 'courseId'], (idx, id) => `${getApiBaseUrl}/xblock/block-v1:${id}`)
+  .attr('type', null, 'sequential')
+  .attr('children', ['id', 'display-name', 'courseId'], (id, name, courseId) => {
+    Factory.reset('topic');
+    return Factory.buildList('topic', 2, null, {
+      topicPrefix: `${id}`, enabledInContext: true, topicNamePrefix: `${name}`, usageKey: `${courseId.replace('course-v1:', 'block-v1:')} +type@vertical+block@vertical_`,
+    });
```
Comment (awais-ansari) [inline src/discussions/in-context-topics/data/__factories__/inContextTopics.factory.js:30]: These lines have a width greater than 120. split these lines into multiple lines.

File: src/discussions/in-context-topics/data/__factories__/inContextTopics.factory.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+Factory.define('topic')
+  .sequence('id', ['topicPrefix'], (idx, topicPrefix) => `${topicPrefix}-${idx}`)
+  .sequence('enabled-in-context', ['enabledInContext'], (idx, enabledInContext) => enabledInContext)
+  .sequence('name', ['topicNamePrefix'], (idx, topicNamePrefix) => `${topicNamePrefix}-${idx}`)
+  .sequence('usage-key', ['usageKey'], (idx, usageKey) => usageKey)
+  .sequence('courseware', ['courseware'], (idx, courseware) => courseware)
+  .attr('thread_counts', () => {
+    Factory.reset('thread-counts');
+    return Factory.build('thread-counts', null, { discussionCount: 1, questionCount: 1 });
+  });
```
Comment (awais-ansari) [inline src/discussions/in-context-topics/data/__factories__/inContextTopics.factory.js:14]: allow this attribute to get dynamic params for discussion count and question count.

File: src/discussions/in-context-topics/data/__factories__/inContextTopics.factory.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+Factory.define('section')
+  .sequence('block_id', (idx) => `${idx}`)
+  .option('topicPrefix', null, '')
+  .sequence('id', ['topicPrefix'], (idx, topicPrefix) => `${topicPrefix}-topic-${idx}`)
+  .attr('courseware', null, true)
+  .sequence('display-name', (idx) => `Introduction ${idx}`)
+  .option('courseId', null, 'course-v1:edX+DemoX+Demo_Course')
+  .sequence('legacy_web_url', ['id', 'courseId'], (idx, id, courseId) => `${getApiBaseUrl}/courses/${courseId}/jump_to/${courseId.replace('course-v1:', 'block-v1:')}+type@chapter+block@${id}?experience=legacy`)
+  .sequence('lms_web_url', ['id', 'courseId'], (idx, id, courseId) => `${getApiBaseUrl}/courses/${courseId}/jump_to/${courseId.replace('course-v1:', 'block-v1:')}+type@chapter+block@${id}`)
+  .sequence('student_view_url', ['id', 'courseId'], (idx, id, courseId) => `${getApiBaseUrl}/xblock/${courseId.replace('course-v1:', 'block-v1:')}+type@chapter+block@${id}`)
```
Comment (awais-ansari) [inline src/discussions/in-context-topics/data/__factories__/inContextTopics.factory.js:42]: These lines have a width greater than 120. split these lines into multiple lines.

File: src/discussions/in-context-topics/data/__factories__/inContextTopics.factory.js (hunk 1 above)
Comment (awais-ansari) [inline src/discussions/in-context-topics/data/__factories__/inContextTopics.factory.js:60]: These lines have width greater than 120. split these lines into multiple lines.

Comment (awais-ansari) [inline src/discussions/in-context-topics/data/__factories__/inContextTopics.factory.js:60]: pass discussion and question count values from here.

File: src/discussions/in-context-topics/data/api.test.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  });
+
+  afterEach(() => {
+    axiosMock.reset();
+  });
+
+  test('successfully get topics', async () => {
+    axiosMock.onGet(`${courseTopicsApiUrl}${courseId}`)
+      .reply(200, (Factory.buildList('topic', 1, null, {
+        topicPrefix: 'noncourseware-topic', enabledInContext: true, topicNamePrefix: 'general-topic', usageKey: '', courseware: false,
```
Comment (awais-ansari) [inline src/discussions/in-context-topics/data/api.test.js:42]: These lines have width greater than 120. split these lines into multiple lines.

File: src/discussions/in-context-topics/data/redux.test.js [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+  });
+
+  afterEach(() => {
+    axiosMock.reset();
+  });
+
+  async function setupMockData() {
+    axiosMock.onGet(`${courseTopicsApiUrl}${courseId}`)
+      .reply(200, (Factory.buildList('topic', 1, null, {
+        topicPrefix: 'noncourseware-topic', enabledInContext: true, topicNamePrefix: 'general-topic', usageKey: '', courseware: false,
```
Comment (awais-ansari) [inline src/discussions/in-context-topics/data/redux.test.js:43]: These lines have width greater than 120. split these lines into multiple lines.

File: src/discussions/in-context-topics/data/selector.test.jsx [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+  });
+
+  afterEach(() => {
+    axiosMock.reset();
+  });
+
+  async function setupMockData() {
+    axiosMock.onGet(`${courseTopicsApiUrl}${courseId}`)
+      .reply(200, (Factory.buildList('topic', 1, null, {
+        topicPrefix: 'noncourseware-topic', enabledInContext: true, topicNamePrefix: 'general-topic', usageKey: '', courseware: false,
```
Comment (awais-ansari) [inline src/discussions/in-context-topics/data/selector.test.jsx:52]: These lines have width greater than 120. split these lines into multiple lines.

File: src/discussions/topics/data/__factories__/topics.factory.js [hunk 9]
```diff
@@ -14,7 +14,6 @@ Factory.define('category')
   .attr('id', [], null)
   .sequence('name', (idx) => `category-${idx}`)
   .attr('thread_counts', [], {
-    discussion: 0,
```
Comment (awais-ansari) [inline src/discussions/topics/data/__factories__/topics.factory.js:17]: I think this is removed by mistake.

File: src/discussions/in-context-topics/data/selectors.js [hunk 10]
```diff
@@ -14,6 +14,9 @@ export const selectNonCoursewareIds = state => state.inContextTopics.nonCoursewa
 
 export const selectUnits = state => state.inContextTopics.units;
 
+export const selectSections = state => state.inContextTopics.coursewareTopics?.map(x => x.children)
+  ?.flat();
+
```
Comment (awais-ansari) [inline src/discussions/in-context-topics/data/selectors.js:19]: we still need this selector?

File: src/discussions/in-context-topics/data/selector.test.jsx [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+    return state;
+  }
+
+  test('should return topics list', async () => {
+    setupMockData().then((state) => {
+      const topics = selectTopics(state);
+
+      expect(topics)
+        .not
+        .toBeUndefined();
```
Comment (awais-ansari) [inline src/discussions/in-context-topics/data/selector.test.jsx:67]: make it one line

**Review by awais-ansari — CHANGES_REQUESTED** [review]
Just small changes. Other than that LGTM.

**Review by awais-ansari — APPROVED** [review]
LGTM

**Review by muhammadadeeltajamul — APPROVED** [review]

**Review by mehaknasir — APPROVED** [review]

**Pull request merged**